### PR TITLE
gRPC + Event bus 

### PR DESCRIPTION
**Pull request by HNDaG**

### CODE REVIEW

File: internal-api/build.gradle.kts [hunk 1]
```diff
@@ -10,4 +8,23 @@ protobuf {
...
+	}
+
+	generateProtoTasks {
+		all().forEach {
+			it.plugins {
+				create("grpc")
+				create("reactor-grpc")
+			}
+		}
+	}
```
Comment (ajax-gnatiuk-s) [inline internal-api/build.gradle.kts]:
```suggestion
	plugins {
		id("grpc") {
			artifact = "io.grpc:protoc-gen-grpc-java:1.58.0"
		}

		id("reactor-grpc") {
			artifact = "com.salesforce.servicelibs:reactor-grpc:1.2.4"
		}
	}

	generateProtoTasks {
		all().forEach {
			it.plugins {
				create("grpc")
				create("reactor-grpc")
			}
		}
	}
	
```

fix formatting

File: internal-api/src/main/kotlin/com/nikitahohulia/api/internal/v2/usersvc/UserEvent.kt [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+object UserEvent {
+
+    private const val VERSION = "v2"
+
+    private const val SERVICE_NAME = "listening_platform"
+
+    private const val SUBDOMAIN = "user"
+
+    private const val PREFIX = "$VERSION.$SERVICE_NAME"
```
Comment (ajax-gnatiuk-s) [inline internal-api/src/main/kotlin/com/nikitahohulia/api/internal/v2/usersvc/UserEvent.kt]:
```suggestion
    private const val PREFIX = "v2.listening_platform.output.pubsub"
```

File: listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/entity/User.kt [hunk 3]
```diff
@@ -10,3 +11,15 @@ data class User(
     val subscriptions: MutableSet<ObjectId> = mutableSetOf(),
     val publisherId: ObjectId? = null
 )
+
+fun User.toProto(): ProtoUser {
+    val builder = com.nikitahohulia.api.internal.v2.usersvc.commonmodels.user.User.newBuilder()
```
Comment (ajax-gnatiuk-s) [inline listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/entity/User.kt]:
```suggestion
    val builder = ProtoUser.newBuilder()
```

File: listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/entity/User.kt [hunk 4]
```diff
@@ -10,3 +11,15 @@ data class User(
...
+        .setEmail(email)
+        .setPassword(password)
+        .setUsername(username)
+
+    if (id!=null) builder.setId(id.toHexString())
+    if (publisherId!=null) builder.setPublisherId(publisherId.toHexString())
+    subscriptions.forEach { builder.addSubscriptions(it.toHexString()) }
+    return builder.build()
```
Comment (ajax-gnatiuk-s) [inline listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/entity/User.kt]:
```kotlin
return  ProtoUser.newBuilder().also{ protoUser ->
    id?.let { protoUser.id = it.toHexString() }
    publisherId?.let { protoUser.publisherId = publisherId.toHexString() }

    protoUser.email = email
    protoUser.password = password
    protoUser.username = username

    protoUser.addAllSubscriptions(
        subscriptions.map { itt.toHexString() }
    )
}
```

File: listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/kafka/UserKafkaReceiver.kt [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    @PostConstruct
+    fun init() {
+        kafkaReceiverUserUpdatedEvent.receiveAutoAck()
+            .flatMap { fluxRecord ->
+                fluxRecord
+                    .map {
+                        userUpdatedEventService.publishEvent(it.value().user)
+                    }
+            }
+            .subscribeOn(Schedulers.boundedElastic())
```
Comment (ajax-gnatiuk-s) [inline listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/kafka/UserKafkaReceiver.kt:25]: why `boundedElastic`?

File: listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/service/PublisherServiceImpl.kt [hunk 6]
```diff
@@ -55,7 +58,8 @@ class PublisherServiceImpl(
             .flatMap { publisher -> publisher.id?.let { userRepository.findByPublisherId(it) } ?: Mono.empty() }
             .flatMap { user ->
                 userRepository.save(user.copy(publisherId = null))
-                    .then(redisUserRepository.update(user.copy(publisherId = null)))
+                    .then(redisUserRepository.update(user.copy(publisherId = null))
+                        .doOnNext { kafkaUserProducer.sendUserUpdatedEventToKafka(it.toProto()) })
```
Comment (ajax-gnatiuk-s) [inline listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/service/PublisherServiceImpl.kt]:
```suggestion
                    .then(
                        redisUserRepository.update(user.copy(publisherId = null))
                    )    
                    .doOnNext { 
                        kafkaUserProducer.sendUserUpdatedEventToKafka(it.toProto()) 
                    }
```

It's easier to read

File: listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/grpc/util/ResponseHelper.kt [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+        return UpdateUserResponse.newBuilder().apply {
+            successBuilder
+                .setUser(user)
+        }.build()
+    }
+
+    fun buildFailureResponseUserUpdated(exception: String, message: String?): UpdateUserResponse {
+        return UpdateUserResponse.newBuilder().apply {
+            failureBuilder
+                .setMessage("UpdateUser failed by $exception: $message")
```
Comment (ajax-gnatiuk-s) [inline listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/grpc/util/ResponseHelper.kt]:
```suggestion
            failureBuilder.message = "UpdateUser failed by $exception: $message"
```

Comment (ajax-gnatiuk-s) [reply listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/grpc/util/ResponseHelper.kt]: check all places in this file

File: listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/grpc/GrpcUserService.kt [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+            .map { devices -> buildSuccessResponseGetAll(devices.map { it.toProto() }) }
+            .onErrorResume { ex ->
+                buildFailureResponseGetAll(
+                    ex.javaClass.simpleName,
+                    ex.message
+                ).toMono()
+            }
+    }
+
+    fun handleGetByUsername(username: String): Flux<GetUserByUsernameResponse> {
```
Comment (ajax-gnatiuk-s) [inline listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/grpc/GrpcUserService.kt]: why this function is not private?

And two functions above

File: listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/service/event/UserUpdatedNatsEventService.kt [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+    override val parser: Parser<UserUpdatedEvent> = UserUpdatedEvent.parser()
+
+    private val dispatcher = connection.createDispatcher()
+
+    override fun subscribeToEvents(userId: String, eventType: String): Flux<UserUpdatedEvent> {
+        return Flux.create { sink ->
+            dispatcher.apply {
+                subscribe(createUserEventNatsSubject(userId, eventType))
+                { message ->
```
Comment (ajax-gnatiuk-s) [inline listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/service/event/UserUpdatedNatsEventService.kt]:
```suggestion
                subscribe(createUserEventNatsSubject(userId, eventType)) { message ->
```

File: listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/kafka/UserKafkaReceiver.kt (hunk 5 above)
Comment (HNDaG) [reply listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/kafka/UserKafkaReceiver.kt:25]: userUpdatedEventService.publishEvent(it.value().user) 
 
 Is a blocking operation, in addition as I understand, consuming messages can also be blocking, so as for me there is no reasons not to use ***boundedElastic***

File: internal-api/src/main/kotlin/com/nikitahohulia/api/internal/v2/usersvc/UserEvent.kt [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+
+object UserEvent {
+
+    private const val SUBDOMAIN = "user"
+
+    private const val PREFIX = "v2.listening_platform.output.pubsub"
+
+    const val UPDATED = "updated"
+
+
```
Comment (ajax-gnatiuk-s) [inline internal-api/src/main/kotlin/com/nikitahohulia/api/internal/v2/usersvc/UserEvent.kt]:
```suggestion
```

File: internal-api/build.gradle.kts (hunk 1 above)
Comment (ajax-gnatiuk-s) [reply internal-api/build.gradle.kts]: Maybe, my previous comment was confusing a bit.

We don't use `tabs` in files, we use `spaces`. In this file you use tabs.

**Review by ajax-gnatiuk-s — APPROVED** [review]

**Pull request merged**